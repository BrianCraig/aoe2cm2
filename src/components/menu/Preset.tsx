import * as React from "react";
import {Link} from "react-router-dom";

class Preset extends React.Component<object, object> {
    public render() {
        return (
            <div>
                <p>Description of a single preset</p>
                <Link to='/'>Go to index</Link>
            </div>
        );
    }
}

export default Preset;