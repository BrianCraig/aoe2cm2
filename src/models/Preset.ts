import Civilisation from "./Civilisation";
import Turn from "./Turn";
import {CivilisationEncoder} from "../util/CivilisationEncoder";
import {Assert} from "../util/Assert";
import Player from "./Player";
import Action from "./Action";
import Exclusivity from "./Exclusivity";

class Preset {

    public static readonly EMPTY: Preset = new Preset('', [], []);

    public static readonly NEW: Preset = new Preset('Change me', Civilisation.ALL, []);

    public static readonly SAMPLE: Preset = new Preset('Default Preset', Civilisation.ALL, [
        Turn.HOST_GLOBAL_BAN,
        Turn.GUEST_GLOBAL_BAN,
        Turn.HOST_HIDDEN_BAN,
        Turn.GUEST_HIDDEN_BAN,
        Turn.REVEAL_ALL,
        Turn.HOST_GLOBAL_PICK,
        Turn.GUEST_GLOBAL_PICK,
        Turn.HOST_PICK,
        Turn.GUEST_PICK,
        Turn.HOST_HIDDEN_PICK,
        Turn.GUEST_HIDDEN_PICK,
        Turn.HOST_HIDDEN_PICK,
        Turn.GUEST_HIDDEN_PICK,
        Turn.HOST_HIDDEN_PICK,
        Turn.GUEST_HIDDEN_PICK,
        Turn.HOST_HIDDEN_PICK,
        Turn.GUEST_HIDDEN_PICK,
        Turn.REVEAL_ALL,
        Turn.HOST_HIDDEN_SNIPE,
        Turn.GUEST_HIDDEN_SNIPE,
        Turn.REVEAL_ALL
    ]);

    public static readonly SIMPLE: Preset = new Preset('Simple Preset', Civilisation.ALL, [
        Turn.HOST_NONEXCLUSIVE_BAN,
        Turn.GUEST_NONEXCLUSIVE_BAN,
        Turn.GUEST_NONEXCLUSIVE_PICK,
        Turn.HOST_NONEXCLUSIVE_PICK
    ]);

    public static readonly ILLEGAL: Preset = new Preset('Illegal Preset', Civilisation.ALL, [
        new Turn(Player.HOST, Action.PICK, Exclusivity.GLOBAL, false, true),
        new Turn(Player.GUEST, Action.PICK, Exclusivity.GLOBAL, false, true),
        new Turn(Player.HOST, Action.PICK, Exclusivity.GLOBAL),
        new Turn(Player.GUEST, Action.PICK, Exclusivity.GLOBAL),
    ]);

    public static readonly SIMPLE_PARALLEL: Preset = new Preset('Simple Parallel', Civilisation.ALL, [
        new Turn(Player.HOST, Action.PICK, Exclusivity.GLOBAL, false, true),
        new Turn(Player.GUEST, Action.PICK, Exclusivity.GLOBAL),
        new Turn(Player.HOST, Action.PICK, Exclusivity.GLOBAL, false, true),
        new Turn(Player.GUEST, Action.PICK, Exclusivity.GLOBAL),
    ]);

    public static readonly REVEAL_TEST: Preset = new Preset('Reveal Test', Civilisation.ALL, [
        new Turn(Player.HOST, Action.PICK, Exclusivity.NONEXCLUSIVE, true),
        new Turn(Player.GUEST, Action.PICK, Exclusivity.NONEXCLUSIVE, true),
        new Turn(Player.HOST, Action.BAN, Exclusivity.NONEXCLUSIVE, true),
        new Turn(Player.GUEST, Action.BAN, Exclusivity.NONEXCLUSIVE, true),
        new Turn(Player.NONE, Action.REVEAL_PICKS, Exclusivity.GLOBAL),
        new Turn(Player.HOST, Action.SNIPE, Exclusivity.NONEXCLUSIVE, true),
        new Turn(Player.GUEST, Action.SNIPE, Exclusivity.NONEXCLUSIVE, true),
        new Turn(Player.HOST, Action.PICK, Exclusivity.NONEXCLUSIVE, true),
        new Turn(Player.GUEST, Action.PICK, Exclusivity.NONEXCLUSIVE, true),
        new Turn(Player.HOST, Action.BAN, Exclusivity.NONEXCLUSIVE, true),
        new Turn(Player.GUEST, Action.BAN, Exclusivity.NONEXCLUSIVE, true),
        new Turn(Player.NONE, Action.REVEAL_PICKS, Exclusivity.GLOBAL),
        new Turn(Player.NONE, Action.REVEAL_BANS, Exclusivity.GLOBAL),
        new Turn(Player.NONE, Action.REVEAL_SNIPES, Exclusivity.GLOBAL),
    ]);

    public static readonly HIDDEN_1V1: Preset = new Preset('Hidden 1v1', Civilisation.ALL, [
        Turn.HOST_HIDDEN_BAN,
        Turn.GUEST_HIDDEN_BAN,
        Turn.GUEST_HIDDEN_BAN,
        Turn.HOST_HIDDEN_BAN,
        Turn.HOST_HIDDEN_BAN,
        Turn.GUEST_HIDDEN_BAN,
        Turn.REVEAL_ALL,
        Turn.HOST_HIDDEN_PICK,
        Turn.GUEST_HIDDEN_PICK,
        Turn.REVEAL_ALL
    ]);

    public static readonly HIDDEN_2V2: Preset = new Preset('Hidden 2v2', Civilisation.ALL, [
        Turn.HOST_HIDDEN_BAN,
        Turn.GUEST_HIDDEN_BAN,
        Turn.HOST_HIDDEN_BAN,
        Turn.GUEST_HIDDEN_BAN,
        Turn.HOST_HIDDEN_BAN,
        Turn.GUEST_HIDDEN_BAN,
        Turn.REVEAL_ALL,
        Turn.HOST_HIDDEN_PICK,
        Turn.GUEST_HIDDEN_PICK,
        Turn.REVEAL_ALL,
        Turn.GUEST_HIDDEN_BAN,
        Turn.HOST_HIDDEN_BAN,
        Turn.REVEAL_ALL,
        Turn.GUEST_HIDDEN_PICK,
        Turn.HOST_HIDDEN_PICK,
        Turn.REVEAL_ALL,
    ]);

    public static readonly HIDDEN_3V3: Preset = new Preset('Hidden 3v3', Civilisation.ALL, [
        Turn.HOST_HIDDEN_BAN,
        Turn.GUEST_HIDDEN_BAN,
        Turn.HOST_HIDDEN_BAN,
        Turn.GUEST_HIDDEN_BAN,
        Turn.REVEAL_ALL,
        Turn.HOST_HIDDEN_PICK,
        Turn.GUEST_HIDDEN_PICK,
        Turn.REVEAL_ALL,
        Turn.HOST_HIDDEN_BAN,
        Turn.GUEST_HIDDEN_BAN,
        Turn.REVEAL_ALL,
        Turn.GUEST_HIDDEN_PICK,
        Turn.HOST_HIDDEN_PICK,
        Turn.REVEAL_ALL,
        Turn.GUEST_HIDDEN_BAN,
        Turn.HOST_HIDDEN_BAN,
        Turn.REVEAL_ALL,
        Turn.HOST_HIDDEN_PICK,
        Turn.GUEST_HIDDEN_PICK,
        Turn.REVEAL_ALL,
    ]);

    public static readonly HIDDEN_4V4: Preset = new Preset('Hidden 4v4', Civilisation.ALL, [
        Turn.HOST_HIDDEN_BAN,
        Turn.GUEST_HIDDEN_BAN,
        Turn.HOST_HIDDEN_BAN,
        Turn.GUEST_HIDDEN_BAN,
        Turn.REVEAL_ALL,
        Turn.HOST_HIDDEN_PICK,
        Turn.GUEST_HIDDEN_PICK,
        Turn.GUEST_HIDDEN_PICK,
        Turn.HOST_HIDDEN_PICK,
        Turn.REVEAL_ALL,
        Turn.HOST_HIDDEN_BAN,
        Turn.GUEST_HIDDEN_BAN,
        Turn.REVEAL_ALL,
        Turn.GUEST_HIDDEN_PICK,
        Turn.HOST_HIDDEN_PICK,
        Turn.REVEAL_ALL,
        Turn.GUEST_HIDDEN_BAN,
        Turn.HOST_HIDDEN_BAN,
        Turn.REVEAL_ALL,
        Turn.GUEST_HIDDEN_PICK,
        Turn.HOST_HIDDEN_PICK,
        Turn.REVEAL_ALL,
    ]);

    public static readonly M_CIVS: Preset = new Preset('M Civs', [Civilisation.MAGYARS, Civilisation.MALAY, Civilisation.MALIANS, Civilisation.MAYANS, Civilisation.MONGOLS], [
        Turn.HOST_PICK, Turn.GUEST_PICK
    ]);

    public readonly name: string;
    public readonly encodedCivilisations: string;
    public readonly turns: Turn[];

    constructor(name: string, civilisations: Civilisation[], turns: Turn[] = []) {
        this.name = name;
        this.encodedCivilisations = CivilisationEncoder.encodeCivilisationArray(civilisations);
        this.turns = turns;
    }

    public static fromPojo(preset: { name: string, encodedCivilisations: string, turns: Turn[] } | undefined): Preset | undefined {
        if (preset === undefined) {
            return undefined;
        }
        Assert.isString(preset.name);
        Assert.isString(preset.encodedCivilisations);
        return new Preset(preset.name, CivilisationEncoder.decodeCivilisationArray(preset.encodedCivilisations), Turn.fromPojoArray(preset.turns));
    }

    public addTurn(turn: Turn) {
        this.turns.push(turn);
    }

    get civilisations(): Civilisation[] {
        return CivilisationEncoder.decodeCivilisationArray(this.encodedCivilisations);
    }
}

export default Preset;