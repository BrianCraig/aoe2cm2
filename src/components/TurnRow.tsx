import * as React from 'react';
import {ComponentElement} from 'react';
import {default as ModelTurn} from "../models/Turn";
import '../pure-min.css'
import '../style2.css'
import Turn from "./Turn";

interface IProps {
    turns: ModelTurn[];
}

class TurnRow extends React.Component<IProps, object> {

    public render() {

        const items: Array<ComponentElement<any, any>> = [];

        for (let i = 0; i < this.props.turns.length; i++) {
            const turn = this.props.turns[i];
            items.push(React.createElement(Turn, {turn, turnNumber: i}));
        }

        return (
            <div className="turn-row">
                <div className="pure-g">
                    <div className="pure-u-1-24 arrow-start">
                        <div>Start</div>
                    </div>

                    {items}

                    <div className="pure-u-1-24 arrow-end">
                        <div><span>End</span></div>
                    </div>
                    <div className="pure-u-1-24" id="firefox-bug">&nbsp;</div>
                </div>
            </div>
        );
    }
}

export default TurnRow;
