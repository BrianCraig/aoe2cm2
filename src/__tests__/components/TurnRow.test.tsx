import {shallow} from "enzyme";
import * as React from "react";
import TurnRow from "../../components/TurnRow";
import Preset from "../../models/Preset";


it('CivGrid renders correctly', () => {
    const component = shallow(<TurnRow turns={Preset.SAMPLE.turns}/>);
    expect(component).toMatchSnapshot();
});
