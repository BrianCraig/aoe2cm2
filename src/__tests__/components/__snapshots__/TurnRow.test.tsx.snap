// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CivGrid renders correctly 1`] = `
<div
  className="turn-row"
>
  <div
    className="pure-g"
  >
    <div
      className="pure-u-1-24 arrow-start"
    >
      <div>
        Start
      </div>
    </div>
    <Turn
      turn={
        Turn {
          "action": "GLOBAL_BAN",
          "player": "HOST",
        }
      }
    />
    <Turn
      turn={
        Turn {
          "action": "GLOBAL_BAN",
          "player": "GUEST",
        }
      }
    />
    <Turn
      turn={
        Turn {
          "action": "HIDDEN_BAN",
          "player": "HOST",
        }
      }
    />
    <Turn
      turn={
        Turn {
          "action": "HIDDEN_BAN",
          "player": "GUEST",
        }
      }
    />
    <Turn
      turn={
        Turn {
          "action": "REVEAL_BANS",
          "player": "NONE",
        }
      }
    />
    <Turn
      turn={
        Turn {
          "action": "HIDDEN_PICK",
          "player": "HOST",
        }
      }
    />
    <Turn
      turn={
        Turn {
          "action": "HIDDEN_PICK",
          "player": "GUEST",
        }
      }
    />
    <Turn
      turn={
        Turn {
          "action": "HIDDEN_PICK",
          "player": "HOST",
        }
      }
    />
    <Turn
      turn={
        Turn {
          "action": "HIDDEN_PICK",
          "player": "GUEST",
        }
      }
    />
    <Turn
      turn={
        Turn {
          "action": "HIDDEN_PICK",
          "player": "HOST",
        }
      }
    />
    <Turn
      turn={
        Turn {
          "action": "HIDDEN_PICK",
          "player": "GUEST",
        }
      }
    />
    <Turn
      turn={
        Turn {
          "action": "HIDDEN_PICK",
          "player": "HOST",
        }
      }
    />
    <Turn
      turn={
        Turn {
          "action": "HIDDEN_PICK",
          "player": "GUEST",
        }
      }
    />
    <Turn
      turn={
        Turn {
          "action": "HIDDEN_PICK",
          "player": "HOST",
        }
      }
    />
    <Turn
      turn={
        Turn {
          "action": "HIDDEN_PICK",
          "player": "GUEST",
        }
      }
    />
    <Turn
      turn={
        Turn {
          "action": "HIDDEN_PICK",
          "player": "HOST",
        }
      }
    />
    <Turn
      turn={
        Turn {
          "action": "HIDDEN_PICK",
          "player": "GUEST",
        }
      }
    />
    <Turn
      turn={
        Turn {
          "action": "REVEAL_PICKS",
          "player": "NONE",
        }
      }
    />
    <Turn
      turn={
        Turn {
          "action": "HIDDEN_SNIPE",
          "player": "HOST",
        }
      }
    />
    <Turn
      turn={
        Turn {
          "action": "HIDDEN_SNIPE",
          "player": "GUEST",
        }
      }
    />
    <Turn
      turn={
        Turn {
          "action": "REVEAL_ALL",
          "player": "NONE",
        }
      }
    />
    <div
      className="pure-u-1-24 arrow-end"
    >
      <div>
        <span>
          End
        </span>
      </div>
    </div>
    <div
      className="pure-u-1-24"
      id="firefox-bug"
    >
      Â 
    </div>
  </div>
</div>
`;
